lib/Dialect/Handshake/HandshakeOps.cpp

```diff
@@ -320,11 +320,46 @@ void ControlMergeOp::build(OpBuilder &builder, OperationState &result,
   result.addAttribute("control", builder.getBoolAttr(true));
 }
 
-// void ControlMergeOp::getCanonicalizationPatterns(OwningRewritePatternList
-// &results,
-//                                           MLIRContext *context) {
-//   results.insert<circt::handshake::EliminateSimpleControlMergesPattern>(context);
-// }
+namespace {
+struct EliminateSimpleControlMergesPattern
+    : mlir::OpRewritePattern<ControlMergeOp> {
+  using mlir::OpRewritePattern<ControlMergeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ControlMergeOp op,
+                                PatternRewriter &rewriter) const override;
+};
+} // namespace
+
+LogicalResult EliminateSimpleControlMergesPattern::matchAndRewrite(
+    ControlMergeOp op, PatternRewriter &rewriter) const {
+  auto dataResult = op.getResult(0);
+  auto choiceResult = op.getResult(1);
+  if (!choiceResult.hasOneUse())
+    return failure();
+
+  auto *choiceUser = choiceResult.getUses().begin().getUser();
+  if (!isa<SinkOp>(choiceUser))
+    return failure();
+
+  auto merge = rewriter.create<MergeOp>(op.getLoc(), dataResult.getType(),
+                                        op.dataOperands());
+
+  for (auto &use : dataResult.getUses()) {
+    auto *user = use.getOwner();
+    rewriter.updateRootInPlace(
+        user, [&]() { user->setOperand(use.getOperandNumber(), merge); });
+  }
```

Comment:
Does `dataResult.replaceAllUsesWith(merge)` work here?

Reply:
It works for the test case I am looking at, but that technically violates the [restrictions](https://mlir.llvm.org/docs/PatternRewriter/#restrictions) of the PatternRewriter. In a more complex graph, it might not be safe to directly use the hook on the operations. Because the ControlMerge doesn't have the same number of results as the Merge, I didn't see a way to use the `rewriter.replace*` hooks.